mars/services/task/execution/ray/shuffle.py

```diff
@@ -0,0 +1,162 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+from typing import List, Iterable
+
+from .....core.operand import ShuffleProxy, MapReduceOperand, OperandStage
+from .....utils import lazy_import
+from ....subtask import Subtask
+
+ray = lazy_import("ray")
+
+
+class ShuffleManager:
+    """Manage shuffle execution for ray by resolve dependencies between mappers outputs and reducers inputs based on
+    mapper and reducer index.
+    """
+
+    def __init__(self, subtask_graph):
+        self.subtask_graph = subtask_graph
+        self._proxy_subtasks = []
+        for subtask in subtask_graph:
+            chunk = subtask.chunk_graph.results[0]
+            if isinstance(chunk.op, ShuffleProxy):
+                self._proxy_subtasks.append(subtask)
+        self.num_shuffles = len(self._proxy_subtasks)
+        self.mapper_output_refs = []
+        self.mapper_indices = {}
+        self.reducer_indices = {}
+        for shuffle_index, proxy_subtask in enumerate(self._proxy_subtasks):
+            # Note that the reducers can also be mappers such as `DuplicateOperand`.
+            mapper_subtasks = subtask_graph.predecessors(proxy_subtask)
+            reducer_subtasks = subtask_graph.successors(proxy_subtask)
+            n_mappers = len(mapper_subtasks)
+            n_reducers = _get_reducer_operand(
+                reducer_subtasks[0].chunk_graph
+            ).n_reducers
+            mapper_output_arr = np.empty((n_mappers, n_reducers), dtype=object)
+            self.mapper_output_refs.append(mapper_output_arr)
+            self.mapper_indices.update(
+                {
+                    subtask: (shuffle_index, mapper_index)
+                    for mapper_index, subtask in enumerate(mapper_subtasks)
+                }
+            )
+            # reducers subtask should be sorted by reducer_index and MapReduceOperand.map should insert shuffle block
+            # in reducers order, otherwise shuffle blocks will be sent to wrong reducers.
+            sorted_filled_reducer_subtasks = self._get_sorted_filled_reducers(
+                reducer_subtasks, n_reducers
+            )
+            self.reducer_indices.update(
+                {
+                    subtask: (shuffle_index, reducer_ordinal)
+                    for reducer_ordinal, subtask in enumerate(
+                        sorted_filled_reducer_subtasks
+                    )
+                }
+            )
+
+    @staticmethod
+    def _get_sorted_filled_reducers(
+        reducer_subtasks: Iterable[Subtask], n_reducers: int
+    ):
+        # For operands such as `PSRSAlign`, sometimes `reducer_subtasks` might be less than `n_reducers`.
+        # fill missing reducers with `None`.
+        filled_reducers = list(range(n_reducers))
```

Comment:
If the `reducer_subtasks` is less than `n_reducers` as the comment, the missing value in the `filled_reducers` will be a number not None. The comment is not match with the code logic.

Reply:
Good catch, Fixed it